### PR TITLE
BB-962: Make base playbook configurable

### PR DESCRIPTION
**Pull request by giovannicimolin**
This PR adds support for changing the name of the base Open edX playbook. This is needed because edX recently changed the name of the base playbook from `edx_sandbox.yml` to `openedx_native.yml`. 
[More info](https://github.com/edx/configuration/pull/5025)

**Testing instructions:**
1. Using `master`, create a new OpenEdXInstance for `ginkgo`, `hawthorn`, `ironwood` and `master` on your Ocim devstack.
2. Pull this branch and apply migrations.
3. Check that all the instances that you created have `OpenEdXInstance.configuration_playbook_name` set to the correct playbook name:
* `ginkgo`: `playbooks/edx_sandbox.yml`
* `hawthorn`: `playbooks/edx_sandbox.yml`
* `ironwood`: `playbooks/openedx_native.yml`
* `master`: `playbooks/openedx_native.yml`
4. Create new instances for each release we support (`ginkgo`, `hawthorn`, `ironwood`, `master`) and leave the `configuration_playbook_name` empty, check if the correct playbook name was filled.
5. Click on **Launch New Appserver** and check if `configuration_playbook_name` was correctly set there too.
6. Check the instances on Ocim stage that were deployed for every type of deployment we support. The instances don't need to provision successfully, you just need to check if the correct playbook was used, and you can do that by checking for this line on the logs:
```
Running playbook "playbooks/openedx_native.yml" from "https://github.com/edx/configuration.git"
```
Deployed instances:
* [Master](https://stage.console.opencraft.com/instance/3280/) 
* [Hawthorn](https://stage.console.opencraft.com/instance/3290/) 
* [Ironwood](https://stage.console.opencraft.com/instance/3300/)
* [Beta test instance ](https://stage.console.opencraft.com/instance/3310/)
* [PR watcher sandbox ](https://stage.console.opencraft.com/instance/3330/)

**Reviewers:**
- [ ] @lgp171188

### CODE REVIEW

File: instance/models/utils.py [hunk 1]
```diff
@@ -79,6 +79,25 @@ def default_setting(name):
     return functools.partial(_get_setting, name)
 
 
+def get_base_playbook_name(openedx_release):
+    """
+    This functions returns the correct main playbook name for the given Open EdX
+    release
+
+    This is needed because OpenEdX changed the main playbook name from
+    `edx_sandbox.yml` to `openedx_native.yml`. This is needed because the old
```
Comment (lgp171188) [inline instance/models/utils.py]:
```suggestion
    `edx_sandbox.yml` to `openedx_native.yml` and the old
```

File: instance/models/utils.py [hunk 2]
```diff
@@ -79,6 +79,25 @@ def default_setting(name):
...
+    releases still use the old playbook name.
```
Comment (lgp171188) [inline instance/models/utils.py]:
```suggestion
    releases before Ironwood still use the old playbook.
```

File: instance/migrations/0114_auto_20190227_1534.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: @giovannicimolin, can the file be named meaningfully instead of using an auto-generated name?

File: instance/migrations/0114_auto_20190227_1534.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    # OpenEdXInstance model
+    queryset = OpenEdXInstance.objects.all()
+    # Make all saves() in a single transaction
+    with transaction.atomic():
+        for item in queryset:
+            item.configuration_playbook_name = get_base_playbook_name(item.openedx_release)
+            item.save()
+
+def backwards_func(apps, schema_editor):
+    pass
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: @giovannicimolin, doesn't it make sense to throw error or at least log a warning when a backward migration is attempted since that will result in loss of data? I know the data for this field can be calculated, but this will still be good to have.

File: instance/migrations/0114_auto_20190227_1534.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2019-02-27 15:34
+from __future__ import unicode_literals
+
+from django.db import migrations, models, transaction
+
+
+def forwards_func(apps, schema_editor):
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: @giovannicimolin, wouldn't it be better to name the function based on what it actually does. Ditto for the reverse data migration function.

File: instance/migrations/0114_auto_20190227_1534.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            'hawthorn',
+        ]
+        if any([release_name in openedx_release for release_name in old_playbook_releases]):
+            return 'playbooks/edx_sandbox.yml'
+        return 'playbooks/openedx_native.yml'
+
+    # We get the model from the versioned app registry;
+    # if we directly import it, it'll be the wrong version
+    OpenEdXInstance = apps.get_model("instance", "OpenEdXInstance")
+    db_alias = schema_editor.connection.alias
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: @giovannicimolin, this is unused.

File: instance/migrations/0114_auto_20190227_1534.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    # We only need to update the OpenEdXInstance, because these AppServer
+    # should be immutable, and new AppServers will use the correct values from
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]:
```suggestion
    # are immutable, and new AppServers will use the correct values from
```

File: instance/migrations/0114_auto_20190227_1534.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        ]
+        if any([release_name in openedx_release for release_name in old_playbook_releases]):
+            return 'playbooks/edx_sandbox.yml'
+        return 'playbooks/openedx_native.yml'
+
+    # We get the model from the versioned app registry;
+    # if we directly import it, it'll be the wrong version
+    OpenEdXInstance = apps.get_model("instance", "OpenEdXInstance")
+    db_alias = schema_editor.connection.alias
+    # We only need to update the OpenEdXInstance, because these AppServer
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: Since the `configuration_playbook_name` is a field in the `OpenEdXInstance` model, I don't think the below explanation is needed. If you think it is still needed, can you check the below suggestions out?
```suggestion
    # We only need to update the OpenEdXInstance, because AppServers
```

File: instance/migrations/0114_auto_20190227_1534.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    # should be immutable, and new AppServers will use the correct values from
+    # OpenEdXInstance model
+    queryset = OpenEdXInstance.objects.all()
+    # Make all saves() in a single transaction
+    with transaction.atomic():
+        for item in queryset:
```
Comment (lgp171188) [inline instance/migrations/0114_auto_20190227_1534.py]: @giovannicimolin, can't you just use the `.update()` method on the queryset using the `F()` method to pass the value to the `get_base_playbook_name()` method?

File: instance/models/openedx_appserver.py [hunk 10]
```diff
@@ -163,6 +167,17 @@ class Meta:
         )
     )
 
+    def get_configuration_playbook_name(self):
```
Comment (lgp171188) [inline instance/models/openedx_appserver.py]: @giovannicimolin, wouldn't it be cleaner to make this a property?

File: instance/tests/integration/integration_instance.py [hunk 11]
```diff
@@ -279,6 +278,7 @@ def test_spawn_appserver(self):
         """
         OpenEdXInstanceFactory(
             name='Integration - test_spawn_appserver',
+            configuration_playbook_name='playbooks/opencraft_integration.yml',
```
Comment (lgp171188) [inline instance/tests/integration/integration_instance.py]: @giovannicimolin, wouldn't it be better to mock the `get_base_playbook_name()` method to return this value for all the tests instead of repeating this argument in every test for every case where we do not expect the playbook to fail?

File: instance/tests/models/test_openedx_ansible_mixins.py [hunk 12]
```diff
@@ -90,7 +90,7 @@ def test_provisioning(
             inventory_str=mock_inventory,
             vars_str=appserver.configuration_settings,
             playbook_path='{}/playbooks'.format(working_dir),
-            playbook_name='edx_sandbox.yml',
+            playbook_name='openedx_native.yml',
```
Comment (lgp171188) [inline instance/tests/models/test_openedx_ansible_mixins.py]: @giovannicimolin, it might be worth testing for the ginkgo, hawthorn releases by passing the release and the expected value via ddt.

File: instance/tests/models/test_utils.py [hunk 13]
```diff
@@ -685,53 +687,53 @@ def test_put_no_prefix(self):
         value = 'value'
         agent.put(key, value)
 
-        _, data = self.client.kv.get(key)
-        fetched_value = data['Value'].decode()
+        _, test_data = self.client.kv.get(key)
```
Comment (lgp171188) [inline instance/tests/models/test_utils.py]: @giovannicimolin, why are these changes needed and how are they related to this PR?

Comment (giovannicimolin) [reply instance/tests/models/test_utils.py]: The name of the variable was `data`, so I kept getting pylint errors because I was overriding `ddt.data`. That's why I've changed these names.

Comment (lgp171188) [reply instance/tests/models/test_utils.py]: @giovannicimolin, we usually just import `ddt` and use it as `ddt.data`, `ddt.unpack` :smiley:

File: instance/migrations/0114_auto_20190227_1534.py (hunk 6 above)
Comment (giovannicimolin) [reply instance/migrations/0114_auto_20190227_1534.py]: ooops, missed that when querying.

File: instance/migrations/0114_auto_20190227_1534.py (hunk 8 above)
Comment (giovannicimolin) [reply instance/migrations/0114_auto_20190227_1534.py]: Makes sense, I'll remove this comment.

File: instance/tests/models/test_utils.py (hunk 13 above)
Comment (giovannicimolin) [reply instance/tests/models/test_utils.py]: I've reverted these changes

File: instance/migrations/0114_auto_20190227_1534.py (hunk 9 above)
Comment (giovannicimolin) [reply instance/migrations/0114_auto_20190227_1534.py]: We can't do this because  `F()` is a operation that is performed on the database, so we can pass this values through any function. Only arithmetic operations are allowed. [ref](https://docs.djangoproject.com/en/2.1/ref/models/expressions/)

File: instance/models/openedx_appserver.py (hunk 10 above)
Comment (giovannicimolin) [reply instance/models/openedx_appserver.py]: Yes, and done.

File: instance/tests/integration/integration_instance.py (hunk 11 above)
Comment (giovannicimolin) [reply instance/tests/integration/integration_instance.py]: Done!

File: instance/migrations/0114_introduce_base_playbook_name.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2019-02-27 15:34
+from __future__ import unicode_literals
+
+import logging
+
+from django.db import migrations, models, transaction
+from django.db.models import F
```
Comment (lgp171188) [inline instance/migrations/0114_introduce_base_playbook_name.py]: @giovannicimolin, this is an unused import

Comment (giovannicimolin) [reply instance/migrations/0114_introduce_base_playbook_name.py]: Done

Comment (giovannicimolin) [reply instance/migrations/0114_introduce_base_playbook_name.py]: Fixed.

**Comment by giovannicimolin** [comment]
@lgp171188 The latest commit is deployed on stage. Do you need me to deploy any other test instances there?

File: cleanup_utils/aws_cleanup.py [hunk 15]
```diff
@@ -129,7 +129,7 @@ def get_user_policy(self, username):
                 UserName=username,
                 PolicyName=DEFAULT_POLICY_NAME
             )
-        except botocore.exceptions.NoSuchEntityException:
+        except self.iam_client.exceptions.NoSuchEntityException:
```
Comment (lgp171188) [inline cleanup_utils/aws_cleanup.py]: @giovannicimolin, didn't we discuss and decide to make a separate PR for this? If so, can you remove it?

File: instance/tests/integration/integration_instance.py [hunk 16]
```diff
@@ -362,7 +361,9 @@ def test_external_databases(self):
         if not settings.DEFAULT_INSTANCE_MYSQL_URL or not settings.DEFAULT_INSTANCE_MONGO_URL:
             print('External databases not configured, skipping integration test')
             return
-        OpenEdXInstanceFactory(name='Integration - test_external_databases')
+        OpenEdXInstanceFactory(
```
Comment (lgp171188) [inline instance/tests/integration/integration_instance.py]: @giovannicimolin , can you revert this and the similar changes to formatting now that we do not require an additional argument to be passed?

File: cleanup_utils/aws_cleanup.py (hunk 15 above)
Comment (giovannicimolin) [reply cleanup_utils/aws_cleanup.py]: Yes, I missed it. I'll update the PR.

File: instance/tests/integration/integration_instance.py (hunk 16 above)
Comment (giovannicimolin) [reply instance/tests/integration/integration_instance.py]: Done!

File: cleanup_utils/aws_cleanup.py (hunk 15 above)
Comment (giovannicimolin) [reply cleanup_utils/aws_cleanup.py]: Done

**Comment by giovannicimolin** [comment]
@lgp171188 I"ve addressed the nits and will merge this as soon as the CI checks are green.
